### PR TITLE
Improved Code and Documentation

### PR DESCRIPTION
**Pull request by felipecosta09**
# Changelog 🎉 

This PR intends to simplify the code so more customers can understand, by making the logic simpler to read and commenting every step that the script does. We make it the code with more functionality with fewer lines and fewer files, so now we have just one script file and one manifest with all the necessary permissions.

We've also removed the dependency of [jq](https://stedolan.github.io/jq/) and specific bash versions, to allow even more customers to get started.

This PR additionally added support for the following:
- Automatically Grant Admin Consent for App Registration.
- The ability for customers to choose their App Registration name.
- Update ```README.md``` to show an end-to-end example on how to run the tool.
- Templates for bug reports and feature requests.

From @felipecosta09 and [Tom Ryan](https://github.com/TomRyan-321)

### CODE REVIEW

File: azcli_conformity.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# Export All Subscriptions ID's to an Array and Count to a Variable
+az account list --refresh > /dev/null
+export SUBSCRIPTIONS=(`az account list | grep id | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`)
+NUM_SUBS=${#SUBSCRIPTIONS[@]}
+
+ # Loop to add each one of the Subscriptions
+ i=0
+ for (( i; i<=$NUM_SUBS; i++ ))
+ do  
+    az role definition create --role-definition '{ \
```
Comment (iaindurham) [inline azcli_conformity.sh:35]: 🔧  This command fails for every subscription except for the first one with the error:
`A role definition cannot be updated with a name that already exists.`

As far as I can see the only ways to get around this are to:
1. create the role once adding all the subscriptions as `AssignableScopes` or 
2. somehow update the `AssignableScopes` after it has been created

The current script takes the first approach (which is where `jq` is used). The second approach would probably be a cleaner solution though I don't know how to do that.

Thoughts?

File: azcli_conformity.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+az ad app permission admin-consent --id $APPID
+
+# Export All Subscriptions ID's to an Array and Count to a Variable
+az account list --refresh > /dev/null
+export SUBSCRIPTIONS=(`az account list | grep id | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`)
+NUM_SUBS=${#SUBSCRIPTIONS[@]}
+
+ # Loop to add each one of the Subscriptions
+ i=0
+ for (( i; i<=$NUM_SUBS; i++ ))
```
Comment (iaindurham) [inline azcli_conformity.sh:33]: 🔧  Currently this loop will iterate n+1 times
```suggestion
 for (( i; i<$NUM_SUBS; i++ ))
```

File: azcli_conformity.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+# Export App Registration ID and Active Directory ID
+export AD_ID=`az account show --query tenantId --output tsv`
+export APPID=`az ad app list --display-name "$APP_NAME" | grep appId | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`
+
+# Create a Service Principal for the App Registration
+az ad sp create --id $APPID > /dev/null
+
+# Grant Admin Consent to application
+az ad app permission admin-consent --id $APPID
```
Comment (iaindurham) [inline azcli_conformity.sh:24]: ❓ Is it reasonable to expect that person that will run this script will always be an Admin?

File: azcli_conformity.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+# Export All Subscriptions ID's to an Array and Count to a Variable
+az account list --refresh > /dev/null
+export SUBSCRIPTIONS=(`az account list | grep id | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`)
+NUM_SUBS=${#SUBSCRIPTIONS[@]}
+
+ # Loop to add each one of the Subscriptions
+ i=0
```
Comment (iaindurham) [inline azcli_conformity.sh:32]: ⛏️ There is a random one space indentation in this block

File: azcli_conformity.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+         "Microsoft.AppConfiguration/configurationStores/ListKeyValue/action", \
+         "Microsoft.Network/networkWatchers/queryFlowLogStatus/action", \
+         "Microsoft.Web/sites/config/list/Action", \
+         "Microsoft.Storage/storageAccounts/queueServices/queues/read" \
+      ], \
+      "DataActions": [], \
+      "NotDataActions": [], \
+      "AssignableScopes": ["/subscriptions/'${SUBSCRIPTIONS[i]}'"] \
+      }' > /dev/null
+   sleep 5
```
Comment (iaindurham) [inline azcli_conformity.sh:48]: ❓ Is this always going to be long enough for the role to be "discoverable"?
While developing the other approach using the ARM script it could take up to 15 seconds to become "discoverable".

File: .editorconfig [hunk 6]
```diff
@@ -1,9 +0,0 @@
-root = true
```
Comment (iaindurham) [inline .editorconfig:1]: ❓ Why did you remove this file? You don't like consistency between different IDEs?

File: README.md [hunk 7]
```diff
@@ -1,73 +1,54 @@
...
-   When running the script against all subscriptions in the Active Directory if there is a subscription which the user running
-the script doesn't have permissions to apply the roles to, the script will fail. Any subscriptions that were processed before
-this subscription will have the roles applied.
+ **PS.: Make sure to use the [Issue Template](https://github.com/cloudconformity/azure-onboarding-scripts/tree/master/.github/ISSUE_TEMPLATE)**
 
-## Contributing
+We :heart: pull requests. If you'd like to fix a bug, contribute to a feature or
+just correct a typo, please feel free to do so.
 
-The code style of the shell script follows the [Google Shell Style Guide](https://google.github.io/styleguide/shellguide.html)
```
Comment (iaindurham) [inline README.md:73]: ❓ You don't think it is a good idea to follow the Google Shell Style Guide? IMO It will make it more maintainable going forward as there will be a standard for consistent code style.

File: README.md [hunk 8]
```diff
@@ -1,73 +1,54 @@
...
 ## Usage
 
-```bash
-bash apply-roles --application-id <App registration client id> [--subscription-id <subscription id>]
+First, you need to install the [Azure CLI Tool](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) and authenticate the tool with your Azure Account, you can check how to do that on the official [Azure CLI Documentation](https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli): 
+
+To use the script, clone the repository to your machine and specify the name of the App Registration desired by executing the shell script file:
+
+
+```html
```
Comment (iaindurham) [inline README.md:21]: ❓ Why `html` and not `shell script`?

```suggestion
```shell script
```

File: azcli_conformity.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+az account list --refresh > /dev/null
+export SUBSCRIPTIONS=(`az account list | grep id | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`)
+NUM_SUBS=${#SUBSCRIPTIONS[@]}
+
+ # Loop to add each one of the Subscriptions
+ i=0
+ for (( i; i<=$NUM_SUBS; i++ ))
+ do  
+    az role definition create --role-definition '{ \
+      "Name": "Custom Role - Cloud One Conformity", \
```
Comment (iaindurham) [inline azcli_conformity.sh:36]: 🔧  The custom role name is hard coded here and on line 49. It would be best if they were put into a variable to remove this duplication to make the code more maintainable.

File: azcli_conformity.sh [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+echo " "
+echo "Hello, Thank you for trying this script out.
+We're configuring your Azure Account to work with Cloud One Conformity, this might take several minutes depending of how many subscriptions you have."
+
+# Create App Registration
+az ad app create --display-name "$APP_NAME" --password $APP_SECRET --credential-description secret --required-resource-accesses @manifest.json > /dev/null
+
+# Export App Registration ID and Active Directory ID
+export AD_ID=`az account show --query tenantId --output tsv`
+export APPID=`az ad app list --display-name "$APP_NAME" | grep appId | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`
```
Comment (iaindurham) [inline azcli_conformity.sh:18]: ⛏️ For constancy with variable naming
```suggestion
export APP_ID=`az ad app list --display-name "$APP_NAME" | grep appId | cut -d ":" -f2 | grep -o '".*"' | sed 's/^"\(.*\)".*/\1/'`
```

File: manifest.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
+[{
```
Comment (iaindurham) [inline manifest.json:1]: ❓ Do you mind explaining what this file is for, please?

File: azcli_conformity.sh [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+      "AssignableScopes": ["/subscriptions/'${SUBSCRIPTIONS[i]}'"] \
+      }' > /dev/null
+   sleep 5
+   az role assignment create --role "Custom Role - Cloud One Conformity" --assignee $APPID --scope /subscriptions/${SUBSCRIPTIONS[i]} > /dev/null
+   sleep 5
+   az role assignment create --role Reader --assignee $APPID --scope /subscriptions/${SUBSCRIPTIONS[i]} > /dev/null
+ done
+
+# Print All the Information
+echo "Here the information that you'll need to use to finish the integration"
```
Comment (iaindurham) [inline azcli_conformity.sh:55]: ⛏️ 
```suggestion
echo "Here is the information that you'll need to use to finish the integration"
```

File: azcli_conformity.sh [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+echo " "
+echo "Which name do you want to give to your App Registration?"
+IFS=$'\n'
+read -r APP_NAME
+echo " "
+echo "Hello, Thank you for trying this script out.
+We're configuring your Azure Account to work with Cloud One Conformity, this might take several minutes depending of how many subscriptions you have."
+
+# Create App Registration
+az ad app create --display-name "$APP_NAME" --password $APP_SECRET --credential-description secret --required-resource-accesses @manifest.json > /dev/null
```
Comment (iaindurham) [inline azcli_conformity.sh:14]: ♻️ Make the description a bit more descriptive?
```suggestion
az ad app create --display-name "$APP_NAME" --password $APP_SECRET --credential-description "Cloud One Conformity secret" --required-resource-accesses @manifest.json > /dev/null
```

Comment (iaindurham) [inline azcli_conformity.sh:14]: ♻️ What about some new lines to make this a bit more readable?
```suggestion
az ad app create \
  --display-name "$APP_NAME" \
  --password $APP_SECRET \
  --credential-description secret \
  --required-resource-accesses @manifest.json > /dev/null
```

Same could be applied to the other long lines?

File: azcli_conformity.sh [hunk 14]
```diff
@@ -0,0 +1,60 @@
+# Define Variables for the Configurations
+export APP_SECRET=$(uuidgen)
```
Comment (iaindurham) [inline azcli_conformity.sh:2]: @brianlamCC Is `uuidgen` an acceptable tool to use for generating a customers secret? This will be used to connect their Azure subscriptions to Conformity.

File: azcli_conformity.sh [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+read -r APP_NAME
+echo " "
+echo "Hello, Thank you for trying this script out.
+We're configuring your Azure Account to work with Cloud One Conformity, this might take several minutes depending of how many subscriptions you have."
+
+# Create App Registration
+az ad app create --display-name "$APP_NAME" --password $APP_SECRET --credential-description secret --required-resource-accesses @manifest.json > /dev/null
+
+# Export App Registration ID and Active Directory ID
+export AD_ID=`az account show --query tenantId --output tsv`
```
Comment (iaindurham) [inline azcli_conformity.sh:17]: ❓ What is the value in `export`ing all of these variables?

File: manifest.json [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+		"resourceAppId": "00000003-0000-0000-c000-000000000000",
+		"resourceAccess": [{
+				"id": "e1fe6dd8-ba31-4d61-89e7-88639da4683d",
```
Comment (iaindurham) [inline manifest.json:4]: ⛏️ This file is indented with tabs where as the bash script is indented with spaces. It would be best if all the files in the repo used the same indentation type.

**Review by iaindurham — COMMENTED** [review]
Along with the review comments I have a few questions:
1. The script doesn't exit early when it encounters and error it just carries on. Is that what you would expect? 
Adding a `set -e` at the top of the script should exit early is any of the `az` commands fail.
2. Would you expect the script to handle if the App registration has already been created? 
Tom has suggested that a customer might like to manually set up a single subscription using the Portal. If this is the case would you expect that they would need to remove that app registration or use a new name when running this script?
3. What would you expect to happen if a customer ran this script more than once?
3. Have you put any thought into how a customer could add the custom and reader role an additional Subscription that was added after this script had been run (the current script doesn't handle this, it was to be the next thing that I tackled)
3. One of the reasons that I went for a script with arguments instead of the "guided" approach was I assumed that customers might want to call this script programatically, do you think that customers wouldn't want to do that?
3. The script has a few [shellcheck](https://github.com/koalaman/shellcheck) warnings. As this is going to be a maintained script it would be good if these could be cleaned up. Unless you have a good reason not to?

_Code reviewed using [code review emoji guide](https://github.com/erikthedeveloper/code-review-emoji-guide)_

File: azcli_conformity.sh (hunk 13 above)
Comment (TomRyan-321) [reply azcli_conformity.sh:14]: Value size restriction, could change to something like "ccsecret"
When present, application key identifier cannot be empty and can be at most 32 bytes.
Parameter name: applicationKeyIdentifierValue

Alternateively I will see if there is a better way of generating the secret that still allows the value to be printed at end of script.

File: azcli_conformity.sh (hunk 3 above)
Comment (iaindurham) [inline azcli_conformity.sh:24]: ❓ When I run the script I get the following error.
```
Forbidden({"ClassName":"Microsoft.Portal.Framework.Exceptions.ClientException",
"Message":"Graph call failed with httpCode=Forbidden, errorCode=Authorization_RequestDenied, errorMessage=This operation can only be performed by an administrator. Sign out and sign in as an administrator or contact one of your organization's administrators., reason=Forbidden, correlationId = 1c3814c2-1d05-4315-9908-4c49811523f0, response = {\"odata.error\":{\"code\":\"Authorization_RequestDenied\",\"message\":{\"lang\":\"en\",\"value\":\"This operation can only be performed by an administrator. Sign out and sign in as an administrator or contact one of your organization's administrators.\"},\"requestId\":\"2b5dff55-205e-4ec6-8e28-3f9581e149f3\",\"date\":\"2020-09-28T09:40:34\"}}","Data":{},"HResult":-2146233088,"XMsServerRequestId":null,"Source":null,"HttpStatusCode":403,"ClientData":{"errorCode":"Authorization_RequestDenied","localizedErrorDetails":{"errorDetail":"This operation can only be performed by an administrator. Sign out and sign in as an administrator or contact one of your organization's administrators."},"operationResults":null,"timeStampUtc":"2020-09-28T09:40:34.4880801Z","clientRequestId":"1c3814c2-1d05-4315-9908-4c49811523f0","internalTransactionId":"1f82a7c1-7e05-4e59-a59a-582ba2717c6f","tenantId":"4e6682d3-a3e7-45be-b8a4-133ac2c6e747","userObjectId":"86fee386-adaf-4d97-a9f9-000ac0a14c8e","exceptionType":"AADGraphException"}})
```

I assumed that it was because I'm not an admin user however these are my permissions:

image

Any ideas as to why this might be the case. Do I need higher permissions?

File: azcli_conformity.sh (hunk 13 above)
Comment (iaindurham) [reply azcli_conformity.sh:14]: Thanks for the explanation I didn't know about those limitations.

I think that `ccsecret` or perhaps `cocsecret` for "Cloud One Conformity" or even `tmsecret` are good options for now.

File: azcli_conformity.sh (hunk 3 above)
Comment (TomRyan-321) [reply azcli_conformity.sh:24]: From microsoft docs:
Granting tenant-wide admin consent requires you to sign in as Global Administrator, an Application Administrator, or a Cloud Application Administrator.

In your case, it appears you should have the required permissions already. One thing that may be potentially stopping is "Admin Consent Workflow (Preview)" if this has been set up then this may even restrict global admins from being able to register, if so we need to build a check for it also understand what impact that has on the overall onboarding. In terms of it being in the script, we are simply mimicking the steps documented as part of the manual setup which assumes they would have that access.

**Comment by iaindurham** [comment]
@felipecosta09 I have created a new branch in this upstream repo called `combined`. I was thinking that if you point this PR to that new branch that it could get merged into that branch which can then be collaborated on by all of us.

What do you think?

File: azcli_conformity.sh (hunk 3 above)
Comment (iaindurham) [reply azcli_conformity.sh:24]: Interesting. The plan is to try out the manual steps you have provided today or tomorrow so will be interesting to see if I can do it on the Portal.

File: manifest.json (hunk 11 above)
Comment (TomRyan-321) [reply manifest.json:1]: The manifest.json contains all the MSGraph and AADGraph permissions required as part of the application registration process. There are alternative ways of building these but this file can be used also from within the Azure Console to set these permissions rather than having the users have to find permissions through the permission selector.

File: azcli_conformity.sh (hunk 9 above)
Comment (TomRyan-321) [reply azcli_conformity.sh:36]: Agree

File: azcli_conformity.sh (hunk 13 above)
Comment (TomRyan-321) [reply azcli_conformity.sh:14]: 1. Agreed this could be handled better, there are more improvements I planned on making but due to my limited Git knowledge didn't want to make additional changes until we worked out what we start from, once we get this into a separate branch making feature branches based on that.
2. Planned on adding in an additional check for existing and update existing app rather than re-creating more
3. Ideally we have improved the logic before that so it gracefully handles multiple executions
4. Planned on adding argument support
5. Agree, also my preferred approach, arguments with all args include default values if not specified for ease of use for first time run, planned on adding arg support for selectively applying to subscriptions etc.
6. Will take a look at these haven't used this before.

File: azcli_conformity.sh (hunk 15 above)
Comment (felipecosta09) [reply azcli_conformity.sh:17]: We need these variables to print at the end of the script so customers using the script can finish the integration in the Conformity console. As soon as the Azure API is released, we don't necessarily need to print these and we can just finish the integration on behalf of the customer by adding the Subscriptions via API.

File: azcli_conformity.sh (hunk 13 above)
Comment (felipecosta09) [reply azcli_conformity.sh:14]: Great add!! definitely better to read this way.

Comment (felipecosta09) [reply azcli_conformity.sh:14]: As Tom said, limitation of Azure... Better ```cocsecret``` if we follow the new nomenclature for Cloud One Conformity, don't you think?

File: azcli_conformity.sh (hunk 4 above)
Comment (felipecosta09) [reply azcli_conformity.sh:32]: We can fix this one...

File: azcli_conformity.sh (hunk 5 above)
Comment (felipecosta09) [reply azcli_conformity.sh:48]: In my tests, 5 seconds work, but I have no problem if we have to increase this number, in the case for some reason does not work in some different customer accounts.

File: azcli_conformity.sh (hunk 3 above)
Comment (felipecosta09) [reply azcli_conformity.sh:24]: Yes because is needed to complete the integration and admin-consent is one of the Conformity requirements to integrate.

File: azcli_conformity.sh (hunk 9 above)
Comment (felipecosta09) [reply azcli_conformity.sh:36]: Agree

File: azcli_conformity.sh (hunk 10 above)
Comment (felipecosta09) [reply azcli_conformity.sh:18]: Agreed

File: azcli_conformity.sh (hunk 12 above)
Comment (felipecosta09) [reply azcli_conformity.sh:55]: Agree

**Comment by felipecosta09** [comment]
That's fine, we can do it this way...

**Comment by iaindurham** [comment]
I'm going to merge this so we can continue to work on the solution.
